src/credentials/CHIPCert.cpp

```diff
@@ -0,0 +1,1008 @@
+ * @param allocFunct     A pointer to a function that should be called to allocate memory buffers to hold
+ *                       intermediate data.
+ * @param freeFunct      A pointer to a function that should be called to release memory buffers allocated
+ *                       by allocFunct.
+ *
+ * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+ **/
+CHIP_ERROR ChipCertificateSet::Init(uint8_t maxCerts, uint16_t decodeBufSize, AllocFunct allocFunct, FreeFunct freeFunct)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Certs = reinterpret_cast<ChipCertificateData *>(allocFunct(sizeof(ChipCertificateData) * maxCerts));
+    VerifyOrExit(Certs != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    CertCount = 0;
+    MaxCerts  = maxCerts;
+
+    mDecodeBuf     = nullptr;
+    mDecodeBufSize = decodeBufSize;
+
+    mAllocFunct = allocFunct;
+    mFreeFunct  = freeFunct;
+
+exit:
+    return err;
+}
+
+/**
+ * @brief Initialize ChipCertificateSet.
+ *        This initialization method is used when all memory stractures needed for operation are
+ *        allocated externally and methods in this class don't need to deal with memory allocations.
+ *
+ * @param certsArray     A pointer to the array of the ChipCertificateData structures.
+ * @param certArraySize  Number of ChipCertificateData entries in the array.
+ * @param decodeBuf      Buffer to use for temporary storage of intermediate processing results.
+ * @param decodeBufSize  Size of decoding buffer.
+ *
+ * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+ **/
+CHIP_ERROR ChipCertificateSet::Init(ChipCertificateData * certsArray, uint8_t certArraySize, uint8_t * decodeBuf,
+                                    uint16_t decodeBufSize)
+{
+    Certs          = certsArray;
+    CertCount      = 0;
+    MaxCerts       = certArraySize;
+    mDecodeBuf     = decodeBuf;
+    mDecodeBufSize = decodeBufSize;
+    mAllocFunct    = nullptr;
+    mFreeFunct     = nullptr;
+    return CHIP_NO_ERROR;
+}
+
+/**
+ * @brief Release resources allocated by this class.
+ **/
+void ChipCertificateSet::Release()
+{
+    if (mFreeFunct != nullptr)
+    {
+        if (Certs != nullptr)
+        {
+            mFreeFunct(Certs);
+            Certs = nullptr;
+        }
+        if (mDecodeBuf != nullptr)
+        {
+            mFreeFunct(mDecodeBuf);
+            mDecodeBuf = nullptr;
+        }
+    }
+}
+
+/**
+ * @brief Clear certificate data loaded into this set.
+ **/
+void ChipCertificateSet::Clear()
+{
+    memset(Certs, 0, sizeof(ChipCertificateData) * MaxCerts);
+    CertCount = 0;
+}
+
+/**
+ * @brief Find certificate in the set.
+ *
+ * @param subjectKeyId  Subject key identifier of the certificate to be found in the set.
+ *
+ * @return A pointer to the certificate data On success. Otherwise, NULL if no certificate found.
+ **/
+ChipCertificateData * ChipCertificateSet::FindCert(const CertificateKeyId & subjectKeyId) const
```

Comment:
const method returns a non-const pointer to something inside the object. Is this safe coding-wise? Should we make the pointer const?